### ISSUE TITLE
Security concern regarding hetero_nn

### ISSUE THREAD

**Issue by YongGuCheng** [FederatedAI/FATE#2915]
Hi All,

On the page "https://github.com/FederatedAI/FATE/tree/master/python/federatedml/nn/hetero_nn",  
**Part III**, in Figure 3  "Backward Propagation of Bottom Model":
Party A obtains the gradient  \delta_{bottom_A}  in plaintext. 
The gradient  \delta_{bottom_A}  will leak the label information of Party B.

Any comments about this point?  Thanks!

**Comment by xutongye** [comment]
Could you explain it in more detail?
As I understand it, \delta_{bottom_A}=\delta_{z_A}*(w_A+noise_acc), where z_A=(w_A+noise_acc) * bottom_A, and \delta_{z_A} is obtained from the backward propagation on guest side. I don't see why the label information is leaked.

**Comment by YongGuCheng** [comment]
> Could you explain it in more detail?
> As I understand it, \delta_{bottom_A}=\delta_{z_A}*(w_A+noise_acc), where z_A=(w_A+noise_acc) * bottom_A, and \delta_{z_A} is obtained from the backward propagation on guest side. I don't see why the label information is leaked.

Thank you for your prompt reply.

As we are looking at the "security/privacy" of the protocol, we may consider a few special cases.

First and foremost, we need to agree upon the fact that \delta_{z_A} (i.e., the gradient of the loss function w.r.t. z_A) may leak the label information of the Guest (Party B). For example, when y_pred  = Sigmoid(z_A),  then \delta_{z_A} surely will leak label information. Even when the top model has multiple layers at the Guest, \delta_{z_A} may still leak label information ( See, e..g, https://arxiv.org/abs/2102.08504).

Second, consider the following two scenarios:
(1) For a special case with w_A = I (the identity matrix), then Party A can easily obtain \delta_{z_A} from \delta_{bottom_A}.
(2) When w_A is not I, and when it is close to convergence in the training process, w_A will stay almost the same between two updating steps for the same x_A, then Party A can recover \delta_{z_A} from the difference between two \delta_{bottom_A}'s for the same x_A.

We are developing other cases and we will post more, if possible.

**Comment by ccchengff** [comment]
> > Could you explain it in more detail?
> > As I understand it, \delta_{bottom_A}=\delta_{z_A}*(w_A+noise_acc), where z_A=(w_A+noise_acc) * bottom_A, and \delta_{z_A} is obtained from the backward propagation on guest side. I don't see why the label information is leaked.
> 
> Thank you for your prompt reply.
> 
> As we are looking at the "security/privacy" of the protocol, we may consider a few special cases.
> 
> First and foremost, we need to agree upon the fact that \delta_{z_A} (i.e., the gradient of the loss function w.r.t. z_A) may leak the label information of the Guest (Party B). For example, when y_pred = Sigmoid(z_A), then \delta_{z_A} surely will leak label information. Even when the top model has multiple layers at the Guest, \delta_{z_A} may still leak label information (See, e.g., https://arxiv.org/abs/2102.08504).
> 
> Second, consider the following two scenarios:
> (1) For a special case with w_A = I (the identity matrix), then Party A can easily obtain \delta_{z_A} from \delta_{bottom_A}.
> (2) When w_A is not I, and when it is close to convergence in the training process, w_A will stay almost the same between two updating steps for the same x_A, then Party A can recover \delta_{z_A} from the difference between two \delta_{bottom_A}'s for the same x_A.
> 
> We are developing other cases and we will post more, if possible.

@xutongye I have the same concern. Let's denote W<sub>A</sub><sup>'</sup> = W<sub>A</sub> + &epsilon;<sub>acc</sub>, then everything will be straightforward.


In the forward propagation,  the output of the interactive layer would be 
z = z<sub>B</sub> +  z<sub>A</sub> + &alpha;<sub>A</sub> * &epsilon;<sub>acc</sub> = &alpha;<sub>B</sub> * W<sub>B</sub> + &alpha;<sub>A</sub> * W<sub>A</sub> + &alpha;<sub>A</sub> * &epsilon;<sub>acc</sub> = &alpha;<sub>B</sub> * W<sub>B</sub> + &alpha;<sub>A</sub> * (W<sub>A</sub> + &epsilon;<sub>acc</sub>) = &alpha;<sub>B</sub> * W<sub>B</sub> + &alpha;<sub>A</sub> * W<sub>A</sub><sup>'</sup>, 
which is a matrix multiplication z = &alpha; * W given &alpha; = concat(&alpha;<sub>A</sub>, &alpha;<sub>B</sub>) and W = concat(W<sub>A</sub><sup>'</sup>, W<sub>B</sub>).


In the backward propagation, Party A eventually obtains 
&delta;<sub>bottom_A</sub> = &delta;<sub>act</sub> * (W<sub>A</sub> + &epsilon;<sub>acc</sub>) = &delta;<sub>act</sub> * W<sub>A</sub><sup>'</sup>.
In fact, this is the backprop of matrix multiplication since
&delta;<sub>act</sub> * W<sub>A</sub><sup>'</sup> = &part;Loss / &part;z * W<sub>A</sub><sup>'</sup> = &part;Loss / &part;&alpha;<sub>A</sub>. 

Consequently, &delta;<sub>bottom_A</sub> is the partial derivative of the loss function w.r.t. &alpha;<sub>A</sub>. According to the existing findings, it may cause label leakage (e.g. <a>https://arxiv.org/abs/2102.08504</a>).

**Comment by xutongye** [comment]
@ccchengff , @YongGuCheng , thanks a lot for the explaination and pointing me to the paper. Yes, I understand the issue after reading the paper. Do you have any idea how we can solve this? I know from the paper that we can add noise to the gradient, but the perfermence of the model will decrease. So, I guess there must be some other ways, where we encrypt all information between parties during the training, and only reveal the model parameters after the training is finitished.

**Comment by xutongye** [comment]
And also, @ccchengff and @YongGuCheng , would you please take a few minutes to have a look at another issue I opened about hetero-NN? https://github.com/FederatedAI/FATE/issues/2928

**Comment by dylan-fan** [comment]
Thanks for your concern on heteroNN implemented in FATE. 

Two different scenarios:
1. In the paper （https://arxiv.org/abs/2102.08504）, there is a hypothesis:  active party(Party B ) just has label y. In their attack experiment, the active party only consists of the last two layers of the deep part.  If more than last two layers， what happens？
2. In the heteroNN， there is a hypothesis:  active party(Party B )  has label y and  xB.  So in this case, it‘s more difficult to leak label y from gradA.

Security to strengthen(Maybe):
1.  In the paper aboved gives some solutions, such as dp ( Trade-off between Utility and Privacy)
2.  The embedding passort technology may be a solution.  please see "Rethinking deep neural network ownership verification: Embedding passports to defeat ambiguity attacks"

If there is a better discovery or a better solution, please let me know and we will optimize it for future releases.

**Comment by YongGuCheng** [comment]
> Thanks for your concern on heteroNN implemented in FATE.
> 
> Two different scenarios:
> 
> 1. In the paper （[https://arxiv.org/abs/2102.08504）](https://arxiv.org/abs/2102.08504%EF%BC%89), there is a hypothesis:  active party(Party B ) just has label y. In their attack experiment, the active party only consists of the last two layers of the deep part.  If more than last two layers， what happens？
> 2. In the heteroNN， there is a hypothesis:  active party(Party B )  has label y and  xB.  So in this case, it‘s more difficult to leak label y from gradA.
> 
> Security to strengthen(Maybe):
> 
> 1. In the paper aboved gives some solutions, such as dp ( Trade-off between Utility and Privacy)
> 2. The embedding passort technology may be a solution.  please see "Rethinking deep neural network ownership verification: Embedding passports to defeat ambiguity attacks"
> 
> If there is a better discovery or a better solution, please let me know and we will optimize it for future releases.


@dylan-fan
